### PR TITLE
Controller and shell revamp

### PR DESCRIPTION
**Pull request by Apollon77**
This PR incudes a completely refactored Controller High Level API and some lower level restructurings. As summary:

* [Added Subscription Timeout detection in Low level API](https://github.com/project-chip/matter.js/commit/a49e6b537c3408fe739b7dc31f417bf617a023b8)
* [Make sure the InteractionClient instance knows/transport it's connected Node-id](https://github.com/project-chip/matter.js/commit/cca5a486ae413c379c980b47df3ebf9a363a6202)
* [Enhance Pase features](https://github.com/project-chip/matter.js/commit/33966e68cfc8674e9bb991e705bf845973e8a4f1)
* [Allow to remove all sessions for a specific node](https://github.com/project-chip/matter.js/commit/7a304fc255dd90a7140f9a8e226d4c8c97a61c1e)
* [Restructure MatterNode base class and move around some methods](https://github.com/project-chip/matter.js/commit/9c9607cff9f74d25330c43d6f64213e03b08e580)
* [Introduce PairedNode class to represent a node paired to our controller](https://github.com/project-chip/matter.js/commit/c519000262c0f219d9b76c7d238279b9aa4b3973) aka High level API for a paired Node!
* [Generalize Discovery and move to own Class](https://github.com/project-chip/matter.js/commit/8bb26d4bc0bf068e764dd0f0f0bd2b4cb86f8acf)
* [make sure to detect error from addNoc call](https://github.com/project-chip/matter.js/commit/cc84b3fb3e684a0bf25eb1231da716b13f072163)
* [Change Commissioning Class to commission a defined node-id](https://github.com/project-chip/matter.js/commit/7eadc30af8aa1982f49c6ccdb4fef3fdf741189e)
* [Restructuring Controller classes](https://github.com/project-chip/matter.js/commit/32fed88b7e1948f4b487e4c70252dcff5951af78)
* Adjust example, tests and such
* [Shell rewrite - be like chip-tool (ok commands not 100% compatible)](https://github.com/project-chip/matter.js/commit/5eab2803611be057ab6504d089c5c4ff0df78672)

See single commits for better context.

This PR adresses:
* https://github.com/orgs/project-chip/projects/11?pane=issue&itemId=36142590
* https://github.com/orgs/project-chip/projects/11/views/1?pane=issue&itemId=36406550
* https://github.com/orgs/project-chip/projects/11/views/1?pane=issue&itemId=34446301 (Partly)

### CODE REVIEW

File: packages/matter.js/src/session/pase/PaseClient.ts [hunk 1]
```diff
@@ -17,6 +18,20 @@
...
 
 export class PaseClient {
+    static async generatePakePasscodeVerifier(setupPinCode: number, pbkdfParameters: PbkdfParameters) {
+        const { w0, L } = await Spake2p.computeW0L(pbkdfParameters, setupPinCode);
+        return ByteArray.concat(ByteArray.from(w0.toArray()), L);
+    }
+
+    static generateRandomPasscode() {
+        let passcode: number;
+        passcode = Math.floor(Math.random() * 99999997 + 1); // prevents 00000000 and 99999999
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/session/pase/PaseClient.ts:28]: @lauckhart WDYT ... leave it and accept or should I change in something like "(Crypto.randombytes(4) % 99999997)+ 1" or such

[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/session/pase/PaseClient.ts:28]: kk will change

[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/session/pase/PaseClient.ts]: adjusted

**Review by lauckhart — APPROVED** [review]
Looks good, I added my usual smattering of helpful comments you can feel free to ignore.  Except maybe on the security stuff and bumping the lib level. ;)

File: packages/matter.js/src/CommissioningController.ts [hunk 2]
```diff
@@ -3,91 +3,112 @@
...
+
+    /**
+     * Controller own Fabric ID used to initialize the Controller the first time and to generate the Root certificate.
+     * Can not be changed afterward.
+     * Default: 1
+     */
+    readonly adminFabricId?: FabricId;
+
+    /**
+     * Fabric Index used to initialize the Controller the first time. Can not be changed afterward.
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/CommissioningController.ts:62]: I think I simply make index static to 1. Already had that though.

File: packages/matter-node-shell.js/src/shell/cmd_config.ts [hunk 3]
```diff
@@ -14,130 +14,290 @@
...
+    },
+) {
+    const { action, wifiSsid, wifiPassword } = args;
+    switch (action) {
+        case "get":
+            console.log(
+                `Current Wifi-Credentials: SSID="${theNode.Store.get<string>(
+                    "WiFiSsid",
+                    "-",
+                )}, Password="${theNode.Store.get<string>("WiFiPassword", "-")}"`,
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter-node-shell.js/src/shell/cmd_config.ts:238]: done

File: packages/matter-node-shell.js/src/shell/cmd_config.ts [hunk 4]
```diff
@@ -14,130 +14,290 @@
...
+                console.log(`Can not change Wi-Fi credentials: New values not provided`);
+                return 1;
+            }
+            theNode.Store.set("WiFiSsid", wifiSsid);
+            theNode.Store.set("WiFiPassword", wifiPassword);
+            console.log(
+                `New Wifi-Credentials: SSID="${theNode.Store.get<string>(
+                    "WiFiSsid",
+                    "-",
+                )}, Password="${theNode.Store.get<string>("WiFiPassword", "-")}"`,
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter-node-shell.js/src/shell/cmd_config.ts:252]: done

File: packages/matter.js/src/CommissioningController.ts [hunk 5]
```diff
@@ -3,91 +3,112 @@
...
 
-    readonly delayedPairing?: boolean;
-    readonly subscribeAllAttributes?: boolean;
-    readonly subscribeMinIntervalFloorSeconds?: number;
-    readonly subscribeMaxIntervalCeilingSeconds?: number;
+    /**
+     * If set to true, the controller will not connect to any device on startup. You need to user connectNode() or
+     * connect() to connect to the relevant nodes.
+     * */
+    readonly delayedConnection?: boolean;
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/CommissioningController.ts:49]: I chose autoConnect, adjusted

File: packages/matter.js/src/device/PairedNode.ts [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+import { logEndpoint } from "./EndpointStructureLogger.js";
+
+const logger = Logger.get("PairedNode");
+
+export type CommissioningControllerNodeOptions = {
+    /**
+     * When set to true all events and attributes are subscribed and value changes are reflected in the ClusterClient
+     * instances. With this reading attributes values is mostly looked up in the locally cached data.
+     */
+    readonly subscribeAllAttributesAndEvents?: boolean;
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/device/PairedNode.ts:63]: Renamed autoSubscribe

File: packages/matter.js/src/device/PairedNode.ts [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+
+    /** Method internally used to ensure that the node is connected by creating a new InteractionClient if needed. */
+    private async ensureConnection() {
+        if (this.interactionClient !== undefined) return this.interactionClient;
+        this.interactionClient = await this.reconnectInteractionClient();
+        return this.interactionClient;
+    }
+
+    /**
+     * Method internally used to initialize the node after the InteractionClient was created and to subscribe
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/device/PairedNode.ts:145]: done

File: packages/matter.js/src/device/PairedNode.ts [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+        >();
+
+        // Add ClusterClients for all server clusters of the device
+        for (const clusterId of descriptorData.serverList) {
+            const cluster = getClusterById(clusterId);
+            const clusterClient = ClusterClient(cluster, endpointId, interactionClient, data[clusterId]);
+            endpointClusters.push(clusterClient);
+        }
+
+        // TODO use the attributes attributeList, acceptedCommands, generatedCommands to crate the ClusterClient/Server objects
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/device/PairedNode.ts:424]: fixed

File: packages/matter.js/src/device/PairedNode.ts [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+            endpointClusters.push(
+                ClusterServer(cluster, /*clusterData.featureMap,*/ clusterData, {}) as ClusterServerObj<
+                    Attributes,
+                    Events
+                >,
+            ); // TODO Add Default handler!
+        }
+
+        if (endpointId === 0) {
+            // Endpoint 0 is the root endpoint, so this object
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/device/PairedNode.ts:439]: fixed

File: packages/matter.js/src/protocol/ControllerDiscovery.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            servers.unshift(lastKnownServer);
+        }
+        const serversSet = new Set(servers);
+
+        const triedServers = new Set<string>();
+        while (true) {
+            logger.debug(
+                `Server addresses to try: ${Array.from(serversSet.values())
+                    .map(server =>
+                        server.type === "udp"
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/protocol/ControllerDiscovery.ts:99]: done

File: packages/matter.js/src/device/PairedNode.ts [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+            }),
+        });
+
+        return {
+            manualPairingCode: ManualPairingCodeCodec.encode({
+                discriminator: discriminator,
+                passcode: passcode,
+            }),
+            qrPairingCode,
+            qrCode: QrCode.encode(qrPairingCode), // TODO: Really export that always?
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/device/PairedNode.ts:609]: changed

**Review by lauckhart — APPROVED** [review]
Looks good!

**Pull request merged**